lightningd/payalgo.c

```diff
@@ -41,14 +143,73 @@ struct pay {
 	 * freed, then allocated at the start of each pay
 	 * attempt to ensure no leaks across long pay attempts */
 	char *try_parent;
+
+	/* Current route being attempted. */
+	struct route_hop *route;
+
+	/* List of failures to pay. */
+	struct list_head pay_failures;
 };
 
+static struct routing_failure *
+dup_routing_failure(const tal_t *ctx, const struct routing_failure *fail)
+{
+	struct routing_failure *nobj = tal(ctx, struct routing_failure);
+
+	nobj->erring_index = fail->erring_index;
+	nobj->failcode = fail->failcode;
+	nobj->erring_node = fail->erring_node;
+	nobj->erring_channel = fail->erring_channel;
+	if (fail->channel_update)
+		nobj->channel_update =  tal_dup_arr(nobj, u8,
+						    fail->channel_update,
+						    tal_count(fail->channel_update),
+						    0);
+	else
+		nobj->channel_update = NULL;
+
+	return nobj;
+}
+
+/* Add a pay_failure from a sendpay_result */
+static void
+add_pay_failure(struct pay *pay,
+		const struct sendpay_result *r)
+{
+	struct pay_failure *f = tal(pay, struct pay_failure);
+
+	/* Append to tail */
+	list_add_tail(&pay->pay_failures, &f->list);
+
+	switch (r->errorcode) {
+	case PAY_UNPARSEABLE_ONION:
+		f->type = FAIL_UNPARSEABLE_ONION;
+		f->onionreply = tal_dup_arr(f, u8, r->onionreply,
+					    tal_count(r->onionreply), 0);
+		break;
+
+	case PAY_TRY_OTHER_ROUTE:
+		f->type = FAIL_PAYMENT_REPLY;
+		f->routing_failure = dup_routing_failure(f,
+							 r->routing_failure);
+		break;
+
+		/* All other errors are disallowed */
+	default:
+		abort();
+	}
+	/* Grab the route */
+	f->route = tal_steal(f, pay->route);
+	pay->route = NULL;
+}
+
 static void
-json_pay_success(struct command *cmd,
+json_pay_success(struct pay *pay,
 		 const struct preimage *payment_preimage,
 		 unsigned int getroute_tries,
 		 unsigned int sendpay_tries)
```

Comment:
I assume getroute_tries and sendpay_tries are the same as in the `struct pay` so these can be removed and maybe further unified with the json_pay_failure case?

Reply:
You are correct, that should be fixed too.